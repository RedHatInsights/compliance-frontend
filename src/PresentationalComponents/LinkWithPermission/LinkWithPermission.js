import React from 'react';
import propTypes from 'prop-types';
import { Tooltip } from '@patternfly/react-core';
import Link from '@redhat-cloud-services/frontend-components/InsightsLink';
import useRoutePermissions from 'Utilities/hooks/useRoutePermissions';

const NoOp = ({ children }) => children;
NoOp.propTypes = {
  children: propTypes.node,
};

/**
 * Component to render a InsightsLink if required permissions are met or render a disabled link with tooltip if not.
 *
 *  @param   {object}             props                  Component props
 *  @param   {string | object}    props.to               ReactRouter to prop, which will be used to determine the required permissions for this path
 *  @param   {React.ReactElement} [props.Component]      Alternative "Link" component to render
 *  @param   {object}             [props.componentProps] Additional props for the Link component
 *  @param   {React.ReactElement} props.children         Component children
 *
 *  @returns {React.ReactElement}                        Component
 *
 *  @category Compliance
 *  @subcategory Components
 *
 */
const LinkWithPermission = ({
  to,
  children,
  Component,
  componentProps = {},
  ...linkProps
}) => {
  const ComponentToRender = Component || Link;
  const { hasAccess, isLoading } = useRoutePermissions(to);
  const hasPermission = !isLoading && hasAccess;
  const TooltipOrDiv = !hasPermission ? Tooltip : NoOp;

  return (
    <TooltipOrDiv
      content={<div>You do not have permissions to perform this action</div>}
    >
      <ComponentToRender
        app="compliance"
        to={to}
        {...componentProps}
        {...linkProps}
        isDisabled={!hasPermission}
      >
        {children}
      </ComponentToRender>
    </TooltipOrDiv>
  );
};

LinkWithPermission.propTypes = {
  to: propTypes.oneOfType([propTypes.string, propTypes.object]),
  children: propTypes.node,
  Component: propTypes.oneOfType([propTypes.func, propTypes.node]),
  componentProps: propTypes.object,
};

export default LinkWithPermission;
