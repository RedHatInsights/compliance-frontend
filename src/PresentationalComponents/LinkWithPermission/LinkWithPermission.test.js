import { render } from '@testing-library/react';
import LinkWithPermission from './LinkWithPermission';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: () => 'Mocked Link',
  useLocation: jest.fn(),
}));

describe('LinkWithPermission', () => {
  it('expect to render without error', () => {
    const { asFragment } = render(
      <LinkWithPermission to="/reports">Test Warning Text</LinkWithPermission>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
