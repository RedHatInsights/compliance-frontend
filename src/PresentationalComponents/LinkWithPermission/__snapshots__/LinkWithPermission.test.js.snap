// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkWithPermission expect to render without error 1`] = `
<DocumentFragment>
  Mocked Link
</DocumentFragment>
`;
